compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -373,11 +413,108 @@ impl<'tcx> SizeSkeleton<'tcx> {
             (SizeSkeleton::Pointer { tail: a, .. }, SizeSkeleton::Pointer { tail: b, .. }) => {
                 a == b
             }
+            // constants are always pre-normalized into a canonical form so this
+            // only needs to check if their pointers are identical.
+            (SizeSkeleton::Generic(a), SizeSkeleton::Generic(b)) => a == b,
             _ => false,
         }
     }
 }
 
+/// To compare constants, we only use this to multiply them, which ensures
+/// that each multiplied term is in sorted order.
+/// i.e. b * a always gets normalized to a * b, in order of the Ord impl
+/// of `Const`.
+/// It will return `None` if multiplication of constants overflows.
+fn mul_sorted_consts<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    a: ty::Const<'tcx>,
+    b: ty::Const<'tcx>,
+) -> Option<ty::Const<'tcx>> {
+    use crate::mir::BinOp::Mul;
+    use ty::ConstKind::Expr;
+    use ty::Expr::Binop;
+
+    // fast-path
+    match (a.kind(), b.kind()) {
+        (Expr(Binop(Mul, ..)), _) | (_, Expr(Binop(Mul, ..))) => {}
+        _ => {
+            let (l, r) = if a < b { (a, b) } else { (b, a) };
+            return Some(tcx.mk_const(Expr(Binop(Mul, l, r)), tcx.types.usize));
+        }
+    }
+
+    let mut buf: SmallVec<[ty::Const<'tcx>; 4]> = SmallVec::new();
+    let mut a = sort_mul(tcx, param_env, a)?;
+    let mut b = sort_mul(tcx, param_env, b)?;
+    while let (Expr(Binop(Mul, al, ar)), Expr(Binop(Mul, bl, br))) = (a.kind(), b.kind()) {
```

Comment:
We should be able to do all of this in one loop instead of having three separate loops.